cart/src/models/repository.js

```diff
@@ -0,0 +1,213 @@
+   * @async
+   * @param {number} cartid - the id associated with a cart
+   */
+  async getCart(cartid) {
+    const carts = this.knex('carts');
+    const checkCart = carts.select('*').where({cartid: cartid});
+    this.logger.debug(`\tQuery: ${checkCart}`);
+
+    const cartsFound = await checkCart;
+    this.logger.debug(`\tResult ${JSON.stringify(cartsFound)}`);
+    return cartsFound;
+  }
+
+  /**
+   * Create a new cart
+   * @async
+   * @param {number} cartid - the id associated with a cart
+   */
+  async createCart(cartid) {
+    const carts = this.knex('carts');
+
+    const cartData = {
+      cartid: cartid,
+      date_created: this.postgresDateStr(),
+      uid: 'user1' // TODO: this shouldn't be hardcoded
+    }
+
+    const createCart = carts.insert(cartData).returning('*');
+    this.logger.debug(`\tQuery: ${createCart}`);
+
+    const created = await createCart;
+    this.logger.debug(`\tResult: ${created}`);
+    return created;
+  }
+
+  /**
+   * Delete a cart
+   * @async
+   * @param {number} cartid - the id associated with a cart
+   */
+  async deleteCart(cartid) {
+    // Remove all products in the cart
+    await this.emptyCart(cartid);
+
+    // Remove the cart itself
+    const carts = this.knex('carts');
+    const query = carts.where({cartid: cartid}).del();
+    this.logger.debug(`\tQuery: ${query}`);
+
+    const removed = await query;
+    this.logger.debug(`\tResult: ${removed}`);
+    return removed;
+  }
+
+  /**
+   * Add a product to the specified cart
+   *
+   * @async
+   * @param {number} cartid - the id associated with a cart
+   * @param {object} product - the product to add, containing at least the fields 'pid' and 'amount_in_cart'
+   */
+  async addProduct(cartid, product) {
+    // Set up the table reference
+    const productsInCart = this.knex(this.resource);
+
+    const productData = {
+      cartid: cartid,
+      amount_in_cart: product.amount_in_cart,
+      pid: product.pid,
+      date_added: this.postgresDateStr()
+    }
+
+    // Then add the product to the cart
+    const addProduct = productsInCart.insert(productData).returning(['pid', 'amount_in_cart']);
+    this.logger.debug(`\tQuery: ${addProduct}`);
+
+    return addProduct;
+  }
+
+  /**
+   * Remove a product from a cart
+   * @async
+   * @param {number} cartid - the id associated with a cart
+   * @param {object} product - the product to remove, containing at least the field 'pid'
+   */
+  async removeProduct(cartid, product) {
+    const productsInCart = this.knex(this.resource);
+
+    const query = productsInCart.where({cartid: cartid, pid: product.pid}).del();
+    this.logger.debug(`\tQuery: ${query}`);
+
+    const nRowsDeleted = await query;
+    this.logger.debug(`\tResult: deleted ${nRowsDeleted} rows.`);
+
+    return nRowsDeleted;
+  }
+
+  /**
+   * Empty a cart (remove all products from it)
+   * @async
+   * @param {number} cartid - the id associated with a cart
+   */
+  async emptyCart(cartid) {
```

Comment:
Except that in the database, the column is `cartid`, because postgres is case-insensitive by default and creates tables in lowercase.

Three options:
1. Make the tables case-sensitive in the database
2. Use lowercase cartid (keep it as it is)
3. Specify `{cartid: cartId}` in the where query, which goes against one of your previous points.

I prefer option two. Which do you (or others) think is best?

Reply:
Except that in the database, the column is cartid, because postgres is case-insensitive by default and creates tables in lowercase.

Three options:

Make the tables case-sensitive in the database
Use lowercase cartid (keep it as it is)
Specify {cartid: cartId} in the where query, which goes against one of your previous points.
I prefer option two. Which do you (or others) think is best?

Reply:
See my other comment below.